File: src/TestContainer/TestContainer.cxx
// Copyright (C) 2007-2016  CEA/DEN, EDF R&D, OPEN CASCADE
//
// Copyright (C) 2003-2007  OPEN CASCADE, EADS/CCR, LIP6, CEA/DEN,
// CEDRAT, EDF R&D, LEG, PRINCIPIA R&D, BUREAU VERITAS
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
//
// See http://www.salome-platform.org/ or email : webmaster.salome@opencascade.com
//

//  SALOME TestContainer : test of container creation and its life cycle
//  File   : TestContainer.cxx
//  Author : Paul RASCLE, EDF - MARC TAJCHMAN, CEA
//  Module : SALOME
//  $Header$
//
#include "utilities.h"
#include <iostream>
#ifndef WIN32
#include <unistd.h>
#endif
#include <SALOMEconfig.h>
#include CORBA_CLIENT_HEADER(SALOME_Component)
#include CORBA_CLIENT_HEADER(SALOME_TestComponent)

#include "SALOME_NamingService.hxx"
#include "NamingService_WaitForServerReadiness.hxx"
#include "Basics_Utils.hxx"
#include "Utils_ORB_INIT.hxx"
#include "Utils_SINGLETON.hxx"
#include "Utils_SALOME_Exception.hxx"
#include "Utils_CommException.hxx"

static std::ostream& operator<<(std::ostream& os, const CORBA::Exception& e)
{
  CORBA::Any tmp;
  tmp<<= e;
  CORBA::TypeCode_var tc = tmp.type();
  const char *p = tc->name();
  os<<"Test blocking exception was catch of the kind : ";
  if ( *p != '\0' ) {
    os<<p;
  } 
  else  { 
    os << tc->id();
  }
  
  return os;
}

Engines::TestComponent_ptr create_instance(Engines::Container_ptr iGenFact,
                                           std::string componenttName)
{
  char* reason;
#if defined(_DEBUG_) || defined(_DEBUG)
  bool isLib =
    iGenFact->load_component_Library(componenttName.c_str(),reason);
  ASSERT(isLib);
#else
  iGenFact->load_component_Library(componenttName.c_str(),reason);
#endif
  CORBA::string_free(reason);
  CORBA::Object_var obj = iGenFact->create_component_instance(componenttName.c_str(), 0);
  Engines::TestComponent_var anInstance = Engines::TestComponent::_narrow(obj);
  MESSAGE("create anInstance");
  SCRUTE(anInstance->instanceName());
  return anInstance._retn();
}

int main (int argc, char * argv[])
{
  // Initializing omniORB
  ORB_INIT &init = *SINGLETON_<ORB_INIT>::Instance() ;
  CORBA::ORB_ptr orb = init( argc , argv ) ;
  //  LocalTraceCollector *myThreadTrace = SALOMETraceCollector::instance(orb);

  try
    {
      SALOME_NamingService _NS(orb) ;
      std::string containerName = "/Containers/" ;
      std::string hostName = Kernel_Utils::GetHostname();
      containerName += hostName + "/FactoryServer";
      NamingService_WaitForServerReadiness(&_NS,containerName);

      CORBA::Object_var obj = _NS.Resolve(containerName.c_str()) ;
      Engines::Container_var iGenFact = Engines::Container::_narrow(obj);
      iGenFact->ping() ;

      int nbInstances = 5;

      std::vector<Engines::TestComponent_var> instances(nbInstances);
    
      MESSAGE("------------------------------- create instances ");
      for (int iter = 0; iter < nbInstances ; iter++)
        {
          instances[iter] = create_instance(iGenFact,"SalomeTestComponent");
        }

      MESSAGE("------------------------------ set env instances ");
      for (int iter = 0; iter < nbInstances ; iter++)
        {
          Engines::TestComponent_var anInstance = instances[iter];
          SCRUTE(anInstance->instanceName());
          Engines::FieldsDict_var dico = new Engines::FieldsDict;
          dico->length(3);
          dico[0].key=CORBA::string_dup("key_0");
          dico[0].value <<="value_0";
          dico[1].key=CORBA::string_dup("key_1");
          dico[1].value <<=(CORBA::UShort)72;
          dico[2].key=CORBA::string_dup("key_2");
          dico[2].value <<=(CORBA::ULong)iter;
          anInstance->setProperties(dico);
          MESSAGE("Coucou " << anInstance->Coucou(iter));
          anInstance->Setenv();
        }

      MESSAGE("---------------------------------- get instances ");
      for (int iter = 0; iter < nbInstances ; iter++)
        {
          Engines::TestComponent_var anInstance = instances[iter];
          SCRUTE(anInstance->instanceName());
          Engines::FieldsDict_var dico2 =  anInstance->getProperties();
          for (CORBA::ULong i=0; i<dico2->length(); i++)
            {
              MESSAGE("dico2["<<i<<"].key="<<dico2[i].key);
              MESSAGE("dico2["<<i<<"].value type ="<<dico2[i].value.type()->kind());
              if (dico2[i].value.type()->kind() == CORBA::tk_string)
                {
                  const char* value;
                  dico2[i].value >>= value;
                  MESSAGE("dico2["<<i<<"].value="<<value);
                }
            }
        }

      MESSAGE("------------------------------- remove instances ");
      for (int iter = 0; iter < nbInstances ; iter++)
        {
          Engines::TestComponent_var anInstance = instances[iter];
          SCRUTE(anInstance->instanceName());
          iGenFact->remove_impl(anInstance) ;
          //iGenFact->finalize_removal() ; // unpredictable results ...
        } 
      MESSAGE("------------------------------- PYTHON ");
      {
//      bool isLib =
//        iGenFact->load_component_Library("SALOME_TestComponentPy");
//      ASSERT(isLib);
//      CORBA::Object_var obj =
//        iGenFact->create_component_instance("SALOME_TestComponentPy",
//                                            0);
//      Engines::TestComponent_var anInstance =
//        Engines::TestComponent::_narrow(obj);
//      MESSAGE("create anInstance");
//      SCRUTE(anInstance->instanceName());
      MESSAGE("------------------------------- create instances ");
      for (int iter = 0; iter < nbInstances ; iter++)
        {
          instances[iter] = create_instance(iGenFact,"SALOME_TestComponentPy");
        }

      MESSAGE("---------------------------------- get instances ");
      for (int iter = 0; iter < nbInstances ; iter++)
        {
          Engines::TestComponent_var anInstance = instances[iter];
          SCRUTE(anInstance->instanceName());
          MESSAGE("Coucou " << anInstance->Coucou(iter));
        }
      }
   
      // Clean-up.
      iGenFact->finalize_removal() ;
      orb->shutdown(0);
    }
  catch(CORBA::COMM_FAILURE& ex) {
    INFOS("Caught system exception COMM_FAILURE -- unable to contact the object.")
      }
  catch(CORBA::SystemException& e) {
    INFOS("Caught a CORBA::SystemException."<<e)
      }
  catch(CORBA::Exception& e) {
    INFOS("Caught CORBA::Exception."<<e)
      }
  catch(ServiceUnreachable& e) {
    INFOS("Caught Exception. "<<e)
      }
  catch(...) {
    INFOS("Caught unknown exception.")
      }

  //  delete myThreadTrace;
  return 0;
}

File: src/Container/Test/testcontainer.py
#  -*- coding: iso-8859-1 -*-
# Copyright (C) 2007-2022  CEA/DEN, EDF R&D, OPEN CASCADE
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
#
# See http://www.salome-platform.org/ or email : webmaster.salome@opencascade.com
#

import unittest
from os import getcwd
from Engines import ContainerParameters, ResourceParameters
import SALOME
import salome

from time import sleep

class TestResourceManager(unittest.TestCase):
    def getContainer(self, name):
        rp = ResourceParameters(name="localhost",
                                hostname="localhost",
                                can_launch_batch_jobs=False,
                                can_run_containers=True,
                                OS="Linux",
                                componentList=[],
                                nb_proc=1,
                                mem_mb=1000,
                                cpu_clock=1000,
                                nb_node=1,
                                nb_proc_per_node=1,
                                policy="first",
                                resList=[])
        cp = ContainerParameters(container_name=name,
                                 mode="start",
                                 workingdir=getcwd(),
                                 nb_proc=1,
                                 isMPI=False,
                                 parallelLib="",
                                 resource_params=rp)
        cm = salome.naming_service.Resolve("/ContainerManager")
        return cm.GiveContainer(cp)

    def checkLoads(self, cont, loads):
        self.assertEqual(len(loads), cont.getNumberOfCPUCores())
        for load in loads:
            self.assertTrue(0.0 <= load <= 1.0)

    def test1(self):
        # Check loadOfCPUCores
        cont = self.getContainer("test_container_1")
        loads1 = cont.loadOfCPUCores()
        self.checkLoads(cont, loads1)
        sleep(1)
        loads2 = cont.loadOfCPUCores()
        self.checkLoads(cont, loads2)
        self.assertNotEqual(loads1, loads2)
        cont.Shutdown()

    def test2(self):
        # Check custom script
        cont = self.getContainer("test_container_2")
        import multiprocessing as mp
        ref_load = [min(0.1*(i+1),1.0) for i in range(mp.cpu_count())]
        cont.setPyScriptForCPULoad('cpu_loads = {}'.format(ref_load))
        loads1 = cont.loadOfCPUCores()
        self.assertEqual(loads1, ref_load)
        cont.resetScriptForCPULoad()
        loads2 = cont.loadOfCPUCores()
        self.checkLoads(cont, loads2)
        cont.Shutdown()

    def test3(self):
        # Check bad script
        cont = self.getContainer("test_container_3")
        cont.setPyScriptForCPULoad("bla-bla-bla")
        self.assertRaises(Exception, cont.loadOfCPUCores)
        cont.Shutdown()

    def test4(self):
        # check memory sizes
        cont = self.getContainer("test_container_4")
        memory_total = cont.getTotalPhysicalMemory()
        memory_in_use = cont.getTotalPhysicalMemoryInUse()
        memory_by_me = cont.getTotalPhysicalMemoryInUseByMe()
        self.assertGreater(memory_total, memory_in_use)
        self.assertGreater(memory_in_use, memory_by_me)
        cont.Shutdown()
    
    def test5(self):
        """
        Test checking memory consumption of container
        """
        cont = self.getContainer("test_container_5")
        memory_by_me_start = cont.getTotalPhysicalMemoryInUseByMe()
        import pickle
        psn = cont.createPyScriptNode("n","""b = bytearray(10485760)""")# create 10MB byte array abroad
        psn.execute([],pickle.dumps(((),{})))
        memory_by_me_end = cont.getTotalPhysicalMemoryInUseByMe()
        self.assertGreater(memory_by_me_end,memory_by_me_start)
        self.assertIn(memory_by_me_end-memory_by_me_start,[10,11,12])# test elevation of memory
        cont.Shutdown()

if __name__ == '__main__':
    salome.standalone()
    salome.salome_init()
    unittest.main()
